neps/nep-0605-sharded-contracts.md

```diff
@@ -0,0 +1,695 @@
+    // Avoid function call access keys from calling this method by requiring
+    // one yocto near as attached balance.
+    // This also stops limited sharded contracts from using this method.
+    near_sdk::assert_one_yocto();
+
+    // Only the actual owner of the tokens should be allowed to call this
+    // function.
+    let my_account_id = HostFunctions::current_account_id();
+    let msg_sender = HostFunctions::predecessor_account_id();
+    assert_eq!(my_account_id, msg_sender);
+
+
+    // Update the account balance
+    let mut my_balance = HostFunctions::storage_read("balance");
+    assert!(my_balance >= amount);
+    my_balance -= amount;
+    HostFunctions::storage_write("balance", my_balance);
+
+    // Call the receiver's account to receive the tokens.
+    let sharded_context = HostFunctions::current_context();
+    match sharded_context {
+        ContractContext::Root => panic!("sharded contract should not run in root"),
+        ContractContext::Sharded { .. } => {
+            near_sdk::Promise::new(receiver)
+                .switch_context(sharded_context)
+                .function_call("receive_tokens", amount, 0, 5 * TGAS);
+        }
+    }
+
+    // This pseudocode assumes that `receive_tokens()` always succeeds.  In a
+    // more complete version, if `receive_tokens()` fails, then this function
+    // should undo the balance decrement above to ensure that tokens are not
+    // lost.
+}
+
+// A smart contract function used to receive FT from a transfer.
+fn receive_tokens(amount: Balance) {
+    // The receiver can only accept tokens from the sender if the sender is
+    // using the same sharded contract code.  Otherwise, a malicious actor might
+    // trick the receiver into minting tokens.
+    //
+    // To implement this check, the receiver has to make sure that the caller of
+    // the function is an instance of the same sharded contract code.
+    let my_context = HostFunctions::current_context();
+    // the unwrap here allows ensures that this function is only called by a sharded contract code.
+    let sender_context = HostFunctions::predecessor_context().unwrap();
+
+    assert_eq!(
+        my_context,
+        sender_context,
+        "Predecessor must be the same sharded contract"
+    );
+
+    // Update the account balance
+    let mut my_balance = HostFunctions::storage_read("balance");
+    my_balance += amount;
+    HostFunctions::storage_write("balance", my_balance);
+
+    // A complete implementation would also emit JSON events here.
+}
+```
+
+With the above in place, following is the flow for how a sharded FT contract would be used.
+
+1. The owner of `ft.near` uses `DeployGlobalContractAction` to deploy the code under the `ft.near` name.
+2. Before `alice.near` can use the sharded FT contract `ft.near`, she has to enable it on her account.
+    - (`receiver_id="alice.near"`)
+    - `SwitchContextAction(Sharded("ft.near"))`
+    - `UseGlobalContractAction("ft.near")`
```

Comment:
Is this action necessary? Since it is only possible to use the global contract code for `ft.near` in the `ft.near` context, it seems redundant.

Reply:
updated just now